### ISSUE TITLE
Funcotator with gnomAD enabled crashes with Bad Request

### ISSUE THREAD

**Issue by alanhoyle** [broadinstitute/gatk#6926]
### Affected tool(s) or class(es)
_Funcotator_

### Affected version(s)
- [ ] Latest public release version GATK 4.1.9.0 with data version funcotator_dataSources.v1.7.20200521g

#### Steps to reproduce

I'm trying to run GATK Funcotator using the funcotator_dataSources.v1.7.20200521g data download.  The command line that I'm using is:
```
gatk Funcotator         \
     --variant cohort.vcf.gz     \
     --reference GRCh38.d1.vd1/GRCh38.d1.vd1.fa         \
     --ref-version hg38   \
     --data-sources-path  funcotator_dataSources.v1.7.20200521g         \
     --output cohort.funcotator.vcf.gz      \
     --output-file-format VCF
```

If I run that command line without the `gnomad_*.tar.gz`'s expanded, it works fine and annotates my `cohort.vcf.gz` into `cohort.funcotator.vcf.gz`.  

Following the directions at [Funcotator Information and Tutorial - 1.1.2.2.1: enabling gnomAD](https://gatk.broadinstitute.org/hc/en-us/articles/360035889931-Funcotator-Information-and-Tutorial#1.1.2.2.1), if I expand both `gnomAD_exome.tar.gz` and `gnomAD_genome.tar.gz`, funcotator dies at startup with a `400 Bad Request` error.  This also happens if I expand either one of the `gnomad_*.tar.gz` files individually.  

#### Expected behavior
Funcotator annotates my VCF and includes gnomAD annotations in the output VCF. 

#### Actual behavior

Crash with 400 Bad Request:

```
Using GATK jar /opt/conda/share/gatk4-4.1.9.0-0/gatk-package-4.1.9.0-local.jar
Running:
    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar /opt/conda/share/gatk4-4.1.9.0-0/gatk-package-4.1.9.0-local.jar Funcotator --variant cohort.vcf.gz --reference GRCh38.d1.vd1/GRCh38.d1.vd1.fa --ref-version hg38 --data-sources-path funcotator_dataSources.v1.7.20200521g --output cohort.funcotator.vcf.gz --output-file-format VCF
14:24:33.589 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/opt/conda/share/gatk4-4.1.9.0-0/gatk-package-4.1.9.0-local.jar!/com/intel/gkl/native/libgkl_compression.so
14:24:33.842 INFO  Funcotator - ------------------------------------------------------------
14:24:33.842 INFO  Funcotator - The Genome Analysis Toolkit (GATK) v4.1.9.0
14:24:33.842 INFO  Funcotator - For support and documentation go to https://software.broadinstitute.org/gatk/
14:24:33.843 INFO  Funcotator - Executing as alanh@r820-2-0.local on Linux v3.10.0-1062.el7.x86_64 amd64
14:24:33.843 INFO  Funcotator - Java runtime: OpenJDK 64-Bit Server VM v1.8.0_265-b11
14:24:33.843 INFO  Funcotator - Start Date/Time: October 29, 2020 2:24:33 PM UTC
14:24:33.843 INFO  Funcotator - ------------------------------------------------------------
14:24:33.843 INFO  Funcotator - ------------------------------------------------------------
14:24:33.844 INFO  Funcotator - HTSJDK Version: 2.23.0
14:24:33.844 INFO  Funcotator - Picard Version: 2.23.3
14:24:33.844 INFO  Funcotator - HTSJDK Defaults.COMPRESSION_LEVEL : 2
14:24:33.844 INFO  Funcotator - HTSJDK Defaults.USE_ASYNC_IO_READ_FOR_SAMTOOLS : false
14:24:33.844 INFO  Funcotator - HTSJDK Defaults.USE_ASYNC_IO_WRITE_FOR_SAMTOOLS : true
14:24:33.844 INFO  Funcotator - HTSJDK Defaults.USE_ASYNC_IO_WRITE_FOR_TRIBBLE : false
14:24:33.844 INFO  Funcotator - Deflater: IntelDeflater
14:24:33.844 INFO  Funcotator - Inflater: IntelInflater
14:24:33.844 INFO  Funcotator - GCS max retries/reopens: 20
14:24:33.844 INFO  Funcotator - Requester pays: disabled
14:24:33.844 INFO  Funcotator - Initializing engine
14:24:34.321 INFO  FeatureManager - Using codec VCFCodec to read file file:///datastore/nextgenout5/share/labs/bioinformatics/alanh/test/nf-germline-exome_sim4/work/3f/5c862c695472a59dfab47a87afe4f3/cohort.vcf.gz
14:24:34.510 INFO  Funcotator - Done initializing engine
14:24:34.511 INFO  Funcotator - Validating sequence dictionaries...
14:24:34.550 INFO  Funcotator - Processing user transcripts/defaults/overrides...
14:24:34.551 INFO  Funcotator - Initializing data sources...
14:24:34.554 INFO  DataSourceUtils - Initializing data sources from directory: funcotator_dataSources.v1.7.20200521g
14:24:34.560 INFO  DataSourceUtils - Data sources version: 1.7.2020521g
14:24:34.560 INFO  DataSourceUtils - Data sources source: ftp://gsapubftp-anonymous@ftp.broadinstitute.org/bundle/funcotator/funcotator_dataSources.v1.7.20200521g.tar.gz
14:24:34.561 INFO  DataSourceUtils - Data sources alternate source: gs://broad-public-datasets/funcotator/funcotator_dataSources.v1.7.20200521.tar.gz
14:24:34.587 INFO  DataSourceUtils - Resolved data source file path: file:///datastore/nextgenout5/share/labs/bioinformatics/alanh/test/nf-germline-exome_sim4/work/3f/5c862c695472a59dfab47a87afe4f3/gencode.v34.annotation.REORDERED.gtf -> file:///datastore/nextgenout5/share/labs/bioinformatics/alanh/test/nf-germline-exome_sim4/work/3f/5c862c695472a59dfab47a87afe4f3/funcotator_dataSources.v1.7.20200521g/gencode/hg38/gencode.v34.annotation.REORDERED.gtf
14:24:34.591 INFO  DataSourceUtils - Resolved data source file path: file:///datastore/nextgenout5/share/labs/bioinformatics/alanh/test/nf-germline-exome_sim4/work/3f/5c862c695472a59dfab47a87afe4f3/gencode.v34.pc_transcripts.fa -> file:///datastore/nextgenout5/share/labs/bioinformatics/alanh/test/nf-germline-exome_sim4/work/3f/5c862c695472a59dfab47a87afe4f3/funcotator_dataSources.v1.7.20200521g/gencode/hg38/gencode.v34.pc_transcripts.fa
14:24:34.599 INFO  DataSourceUtils - Resolved data source file path: file:///datastore/nextgenout5/share/labs/bioinformatics/alanh/test/nf-germline-exome_sim4/work/3f/5c862c695472a59dfab47a87afe4f3/acmg59_test_cleaned.txt -> file:///datastore/nextgenout5/share/labs/bioinformatics/alanh/test/nf-germline-exome_sim4/work/3f/5c862c695472a59dfab47a87afe4f3/funcotator_dataSources.v1.7.20200521g/acmg_rec/hg38/acmg59_test_cleaned.txt
14:24:34.608 INFO  DataSourceUtils - Resolved data source file path: file:///datastore/nextgenout5/share/labs/bioinformatics/alanh/test/nf-germline-exome_sim4/work/3f/5c862c695472a59dfab47a87afe4f3/clinvar_20180429_hg38.vcf -> file:///datastore/nextgenout5/share/labs/bioinformatics/alanh/test/nf-germline-exome_sim4/work/3f/5c862c695472a59dfab47a87afe4f3/funcotator_dataSources.v1.7.20200521g/clinvar/hg38/clinvar_20180429_hg38.vcf
14:24:34.614 INFO  DataSourceUtils - Resolved data source file path: file:///datastore/nextgenout5/share/labs/bioinformatics/alanh/test/nf-germline-exome_sim4/work/3f/5c862c695472a59dfab47a87afe4f3/LMM_Path_LP_VUS5-variants-6-12-18.sorted_liftover_b38.corrected.vcf -> file:///datastore/nextgenout5/share/labs/bioinformatics/alanh/test/nf-germline-exome_sim4/work/3f/5c862c695472a59dfab47a87afe4f3/funcotator_dataSources.v1.7.20200521g/lmm_known/hg38/LMM_Path_LP_VUS5-variants-6-12-18.sorted_liftover_b38.corrected.vcf
14:24:34.617 INFO  DataSourceUtils - Resolved data source file path: file:///datastore/nextgenout5/share/labs/bioinformatics/alanh/test/nf-germline-exome_sim4/work/3f/5c862c695472a59dfab47a87afe4f3/acmg_lof.tsv -> file:///datastore/nextgenout5/share/labs/bioinformatics/alanh/test/nf-germline-exome_sim4/work/3f/5c862c695472a59dfab47a87afe4f3/funcotator_dataSources.v1.7.20200521g/acmg_lof/hg38/acmg_lof.tsv
14:24:35.311 INFO  Funcotator - Shutting down engine
[October 29, 2020 2:24:35 PM UTC] org.broadinstitute.hellbender.tools.funcotator.Funcotator done. Elapsed time: 0.03 minutes.
Runtime.totalMemory()=2055733248
code:      400
message:   400 Bad Request
{
  "error": "invalid_grant",
  "error_description": "Bad Request"
}
reason:    null
location:  null
retryable: false
com.google.cloud.storage.StorageException: 400 Bad Request
{
  "error": "invalid_grant",
  "error_description": "Bad Request"
}
	at com.google.cloud.storage.spi.v1.HttpStorageRpc.translate(HttpStorageRpc.java:229)
	at com.google.cloud.storage.spi.v1.HttpStorageRpc.get(HttpStorageRpc.java:439)
	at com.google.cloud.storage.StorageImpl$5.call(StorageImpl.java:244)
	at com.google.cloud.storage.StorageImpl$5.call(StorageImpl.java:241)
	at shaded.cloud_nio.com.google.api.gax.retrying.DirectRetryingExecutor.submit(DirectRetryingExecutor.java:105)
	at shaded.cloud_nio.com.google.cloud.RetryHelper.run(RetryHelper.java:76)
	at shaded.cloud_nio.com.google.cloud.RetryHelper.runWithRetries(RetryHelper.java:50)
	at com.google.cloud.storage.StorageImpl.get(StorageImpl.java:240)
	at com.google.cloud.storage.contrib.nio.CloudStorageFileSystemProvider.checkAccess(CloudStorageFileSystemProvider.java:736)
	at java.nio.file.Files.exists(Files.java:2385)
	at org.broadinstitute.hellbender.tools.funcotator.dataSources.DataSourceUtils.assertPathFilePropertiesField(DataSourceUtils.java:938)
	at org.broadinstitute.hellbender.tools.funcotator.dataSources.DataSourceUtils.assertConfigFilePropertiesAreValid(DataSourceUtils.java:841)
	at org.broadinstitute.hellbender.tools.funcotator.dataSources.DataSourceUtils.getAndValidateDataSourcesFromPaths(DataSourceUtils.java:216)
	at org.broadinstitute.hellbender.tools.funcotator.Funcotator.onTraversalStart(Funcotator.java:776)
	at org.broadinstitute.hellbender.engine.GATKTool.doWork(GATKTool.java:1047)
	at org.broadinstitute.hellbender.cmdline.CommandLineProgram.runTool(CommandLineProgram.java:140)
	at org.broadinstitute.hellbender.cmdline.CommandLineProgram.instanceMainPostParseArgs(CommandLineProgram.java:192)
	at org.broadinstitute.hellbender.cmdline.CommandLineProgram.instanceMain(CommandLineProgram.java:211)
	at org.broadinstitute.hellbender.Main.runCommandLineProgram(Main.java:160)
	at org.broadinstitute.hellbender.Main.mainEntry(Main.java:203)
	at org.broadinstitute.hellbender.Main.main(Main.java:289)
Caused by: shaded.cloud_nio.com.google.api.client.http.HttpResponseException: 400 Bad Request
{
  "error": "invalid_grant",
  "error_description": "Bad Request"
}
	at shaded.cloud_nio.com.google.api.client.http.HttpRequest.execute(HttpRequest.java:1097)
	at shaded.cloud_nio.com.google.auth.oauth2.UserCredentials.refreshAccessToken(UserCredentials.java:197)
	at shaded.cloud_nio.com.google.auth.oauth2.OAuth2Credentials.refresh(OAuth2Credentials.java:157)
	at shaded.cloud_nio.com.google.auth.oauth2.OAuth2Credentials.getRequestMetadata(OAuth2Credentials.java:145)
	at shaded.cloud_nio.com.google.auth.http.HttpCredentialsAdapter.initialize(HttpCredentialsAdapter.java:91)
	at shaded.cloud_nio.com.google.cloud.http.HttpTransportOptions$1.initialize(HttpTransportOptions.java:159)
	at shaded.cloud_nio.com.google.cloud.http.CensusHttpModule$CensusHttpRequestInitializer.initialize(CensusHttpModule.java:109)
	at shaded.cloud_nio.com.google.api.client.http.HttpRequestFactory.buildRequest(HttpRequestFactory.java:88)
	at shaded.cloud_nio.com.google.api.client.googleapis.services.AbstractGoogleClientRequest.buildHttpRequest(AbstractGoogleClientRequest.java:422)
	at shaded.cloud_nio.com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:541)
	at shaded.cloud_nio.com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:474)
	at shaded.cloud_nio.com.google.api.client.googleapis.services.AbstractGoogleClientRequest.execute(AbstractGoogleClientRequest.java:591)
	at com.google.cloud.storage.spi.v1.HttpStorageRpc.get(HttpStorageRpc.java:436)
	... 19 more
```

**Comment by alanhoyle** [comment]
As an aside, I downloaded the gnomAD VCFs locally using gsutil, modified the `gnomAD_exome.config` to refer to them and it works:

```
$ cat gnomAD_exome.config
name = gnomAD_exome
version = 2.1
src_file = gnomad.exomes.r2.1.sites.liftoverToHg38.INFO_ANNOTATIONS_FIXED.vcf.gz
# src_file = gs://broad-public-datasets/funcotator/gnomAD_2.1_VCF_INFO_AF_Only/hg38/gnomad.exomes.r2.1.sites.liftoverToHg38.INFO_ANNOTATIONS_FIXED.vcf.gz
# [...]
```

**Comment by droazen** [comment]
@alanhoyle Can you tell us whether the 400 Bad Request error is repeatable -- did you see it more than once? Oftentimes when accessing cloud data we encounter transient errors like this that go away on their own.

**Comment by alanhoyle** [comment]
It happened multiple times over the course of a couple days.  Since I
downloaded the full gnomad exome data locally, I haven't tested again.

--
  -  Alan Hoyle  -  alan@alanhoyle.com  -  http://www.alanhoyle.com/  -

**Comment by jonn-smith** [comment]
OK, sounds good. Something may have changed out from underneath us as far as the gnomAD data goes.  We'll look into it.

**Comment by jonn-smith** [comment]
I have tested this with a fresh `gcloud` client and have not been able to reproduce the error.

I did find an article from someone else who got the `400: invalid_grant` error: https://blog.timekit.io/google-oauth-invalid-grant-nightmare-and-how-to-fix-it-9f4efaf1da35

The long and short of it is that it's an authentication issue.  Can you verify that the authentication you're using on the terminal is valid?  That is, can you get at other public resources on gcloud?

**Comment by alanhoyle** [comment]
I was able to download the  gnomAD VCFs with a gsutil cp command within a
couple hours of the remote funcotator failures.

--
Alan Hoyle - alan@alanhoyle.com - alanhoyle.com
------------------------------
*From:* Jonn Smith <notifications@github.com>
*Sent:* Tuesday, November 10, 2020 9:59:06 PM
*To:* broadinstitute/gatk <gatk@noreply.github.com>
*Cc:* Alan Hoyle <alan@alanhoyle.com>; Mention <mention@noreply.github.com>
*Subject:* Re: [broadinstitute/gatk] Funcotator with gnomAD enabled crashes
with Bad Request (#6926)


I have tested this with a fresh gcloud client and have not been able to
reproduce the error.

I did find an article from someone else who got the 400: invalid_grant
error:
https://blog.timekit.io/google-oauth-invalid-grant-nightmare-and-how-to-fix-it-9f4efaf1da35
<https://www.google.com/url?q=https://blog.timekit.io/google-oauth-invalid-grant-nightmare-and-how-to-fix-it-9f4efaf1da35&source=gmail-imap&ust=1605668351000000&usg=AOvVaw03ZXI9QiPy1AFI5zfsFIjB>

The long and short of it is that it's an authentication issue. Can you
verify that the authentication you're using on the terminal is valid? That
is, can you get at other public resources on gcloud?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://www.google.com/url?q=https://github.com/broadinstitute/gatk/issues/6926%23issuecomment-725097056&source=gmail-imap&ust=1605668351000000&usg=AOvVaw0sZboplCCqclv3xBdQk3Fb>,
or unsubscribe

**Comment by droazen** [comment]
Closing as resolved.
